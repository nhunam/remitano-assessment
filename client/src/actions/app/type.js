import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  CLEAR_MESSAGE: undefined
});
