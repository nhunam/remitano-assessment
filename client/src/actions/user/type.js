import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  USER_CREATE: undefined,
  USER_LOGIN: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined
});
