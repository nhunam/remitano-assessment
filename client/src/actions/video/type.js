import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  VIDEO_FETCH: undefined,
  VIDEO_FETCH_SUCCESS: undefined,
  VIDEO_FETCH_FAILURE: undefined,
  VIDEO_SHARING: undefined,
  VIDEO_SHARING_SUCCESS: undefined,
  VIDEO_SHARING_FAILURE: undefined
});
