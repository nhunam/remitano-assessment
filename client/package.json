{
  "name": "video-sharing-react",
  "version": "0.1.0",
  "private": true,
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/node_modules/",
      "!<rootDir>/src/containers/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/actions/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/configurations/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/App.jsx",
      "!<rootDir>/src/serviceWorker.js",
      "!<rootDir>/src/index.js"
    ],
    "coverageThreshold": {
      "global": {
        "branches": 90,
        "functions": 90,
        "lines": 90,
        "statements": 90
      }
    }
  },
  "dependencies": {
    "axios": "^0.19.0",
    "bootstrap": "^4.4.1",
    "classnames": "^2.2.6",
    "enzyme": "^3.10.0",
    "enzyme-adapter-react-16": "^1.15.1",
    "flow-bin": "^0.113.0",
    "http-proxy-middleware": "^1.0.6",
    "immutability-helper": "^3.0.1",
    "jwt-decode": "^2.2.0",
    "keymirror": "^0.1.1",
    "node-sass": "^4.13.0",
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    "react-infinite-scroller": "^1.2.4",
    "react-player": "^1.14.2",
    "react-redux": "^7.1.3",
    "react-router": "^5.1.2",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.2.0",
    "react-test-renderer": "^16.12.0",
    "redux": "^4.0.4",
    "redux-actions": "^2.6.5",
    "redux-logger": "^3.0.6",
    "redux-mock-store": "^1.5.3",
    "redux-persist": "^6.0.0",
    "redux-saga": "^1.1.3",
    "serialize-javascript": "2.1.1",
    "styled-components": "^4.4.1"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "flow": "flow"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "redux-saga-test-plan": "^4.0.0-rc.3"
  }
}
