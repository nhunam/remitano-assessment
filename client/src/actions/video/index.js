import { ActionTypes } from "./type";
import { createActions } from "redux-actions";

export const { videoFetch: fetch, videoSharing: shareVideo } = createActions({
  [ActionTypes.VIDEO_FETCH]: param => ({ ...param }),
  [ActionTypes.VIDEO_SHARING]: param => ({ ...param })
});
