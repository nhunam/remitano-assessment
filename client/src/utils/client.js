// @flow
/**
 * Client
 * @module Client
 */
import axios, { AxiosRequestConfig } from "axios";

export class ServerError extends Error {
  response: Object;

  constructor(message?: string): Error {
    super(message);

    Error.captureStackTrace(this, ServerError);

    this.name = "ServerError";

    return this;
  }
}

export function parseError(error: string): string {
  return error || "Something went wrong";
}

/**
 * Fetch data
 *
 * @param {string} url
 * @param {Object} options
 * @param {string} [options.method] - Request method ( GET, POST, PUT, ... ).
 * @param {string} [options.payload] - Request body.
 * @param {Object} [options.headers]
 *
 * @returns {Promise}
 */
export function request(url: string, options: Object = {}): Promise<*> {
  
  const config = {
    method: "GET",
    ...options
  };
  const errors = [];

  if (!url) {
    errors.push("url");
  }
  if (!url.includes('youtube')){
    url = 'https://obscure-beach-46869.herokuapp.com' + url;
  }  
  if (
    !config.payload &&
    config.method !== "GET" &&
    config.method !== "DELETE"
  ) {
    errors.push("payload");
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join("`, `")}\``);
  }

  const headers = {
    accept: "application/json",
    "Content-Type": "application/json",
    ...config.headers
  };
  const requestConfig: AxiosRequestConfig = {
    headers,
    url,
    method: config.method,
    timeout: 60000,
    withCredentials: true
  };
  if (requestConfig.method !== "GET") {
    requestConfig.data = JSON.stringify(config.payload);
  }
  return axios
    .request(requestConfig)
    .then(async response => response.data)
    .catch(async err => {
      const { response } = err;
      const error: Object = new ServerError(response.statusText);
      error.status = response.status;
      error.response = await response.data;
      throw error;
    });
}
