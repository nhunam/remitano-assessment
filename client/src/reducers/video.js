import { ActionTypes } from "actions/video/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const videoState = {
  data: [],
  page: {
    total_pages: 1,
    has_next: false,
    has_previous: false,
    current_page: 0,
    total_elements: 0
  }
};

export default {
  videos: handleActions(
    {
      [ActionTypes.VIDEO_FETCH]: (state, { payload }) =>
        immutable(state, {
          data: { $set: payload.page_index === 0 ? [] : state.data },
          page: {
            $set:
              payload.page_index === 0
                ? {
                    total_pages: 1,
                    has_next: false,
                    has_previous: false,
                    current_page: 0,
                    total_elements: 0
                  }
                : state.page
          }
        }),
      [ActionTypes.VIDEO_FETCH_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          data: { $set: [...state.data, ...payload.data] },
          page: { $set: payload.page }
        }),
      [ActionTypes.VIDEO_FETCH_FAILURE]: (state, { payload }) =>
        immutable(state, {
          status: { $set: AppStatus.READY }
        })
    },
    videoState
  )
};
