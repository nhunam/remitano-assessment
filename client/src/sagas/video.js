import { all, call, put, takeLatest } from "redux-saga/effects";

import { ActionTypes } from "actions/video/type";
import { AppConfig } from "configurations";
import { parseYoutubeId } from "utils/helpers";
import { request } from "utils/client";

/**
 * Fetch Video
 */
export function* fetch(param) {
  try {
    const payload = {
      method: "GET"
    };
    const paramPayload = param.payload;
    const url = `/videos?page_index=${paramPayload.page_index}`;
    const response = yield call(request, url, payload);

    yield put({
      type: ActionTypes.VIDEO_FETCH_SUCCESS,
      payload: response
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.VIDEO_FETCH_FAILURE,
      payload: err
    });
  }
}

/**
 * Fetch Video
 */
export function* share(param) {
  try {
    const { payload } = param;
    const { videoUrl, userToken } = payload;
    const { youtubeAPIKey, youtubeAPIURL } = AppConfig;
    const requestPayload = {
      method: "GET"
    };
    const videoId = parseYoutubeId(videoUrl);
    // const origin = encodeURI(window.location.href);
    const url = `${youtubeAPIURL}?part=snippet&id=${videoId}&key=${youtubeAPIKey}`;
    const response = yield call(request, url, requestPayload);
    const { items } = response;
    const { snippet } = items[0];
    const { title, description } = snippet;

    const sharingPayload = {
      title: title,
      description: description,
      url: videoUrl
    };
    const headers = { Authorization: `Bearer ${userToken}` };
    const requestSharingPayload = {
      method: "POST",
      payload: sharingPayload,
      headers: headers
    };
    yield call(request, "/videos", requestSharingPayload);

    yield put({
      type: ActionTypes.VIDEO_SHARING_SUCCESS
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.VIDEO_SHARING_FAILURE,
      payload: err
    });
  }
}

/**
 * Video Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.VIDEO_FETCH, fetch)]);
  yield all([takeLatest(ActionTypes.VIDEO_SHARING, share)]);
}
