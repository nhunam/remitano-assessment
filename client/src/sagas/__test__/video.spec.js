import * as matchers from "redux-saga-test-plan/matchers";

import video, { fetch, share } from "sagas/video";

import { ActionTypes } from "actions/video/type";
import { expectSaga } from "redux-saga-test-plan";
import { request } from "utils/client";

describe("User Sagas", () => {
  it("should have the expected watchers", done => {
    expectSaga(video)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      });
  });

  it("should match the fetch saga", () => {
    const fakeDataReturn = {
      data: [],
      page: {
        total_pages: 1,
        has_next: false,
        has_previous: false,
        current_page: 0,
        total_elements: 0
      }
    };
    expectSaga(fetch, { payload: { page_index: 1 } })
      .provide([[matchers.call.fn(request), fakeDataReturn]])
      .put({
        type: ActionTypes.VIDEO_FETCH_SUCCESS,
        payload: {
          data: [],
          page: {
            total_pages: 1,
            has_next: false,
            has_previous: false,
            current_page: 0,
            total_elements: 0
          }
        }
      })
      .run();
  });

  it("should match the sharing saga", () => {
    const fakeDataReturn = {
      items: [
        {
          snippet: {
            title: "test",
            description: "test"
          }
        }
      ]
    };
    expectSaga(share, {
      payload: {
        videoUrl: "https://www.youtube.com/watch?v=EGhSY00oCk",
        token: "test"
      }
    })
      .provide([[matchers.call.fn(request), fakeDataReturn]])
      .put({
        type: ActionTypes.VIDEO_SHARING_SUCCESS
      })
      .run();
  });
});
