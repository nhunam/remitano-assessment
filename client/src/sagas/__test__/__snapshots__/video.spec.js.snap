// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Sagas should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "VIDEO_FETCH",
          "@@redux-saga-test-plan/json/function/fetch",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "VIDEO_SHARING",
          "@@redux-saga-test-plan/json/function/share",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "VIDEO_FETCH",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "VIDEO_SHARING",
      },
      "type": "TAKE",
    },
  ],
}
`;
